lib/msal-browser/src/cache/BrowserCacheManager.ts

```diff
@@ -0,0 +1,604 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Constants, PersistentCacheKeys, StringUtils, AuthorizationCodeRequest, ICrypto, AccountEntity, IdTokenEntity, AccessTokenEntity, RefreshTokenEntity, AppMetadataEntity, CacheManager, ServerTelemetryEntity, ThrottlingEntity, ProtocolUtils } from "@azure/msal-common";
+import { CacheOptions } from "../config/Configuration";
+import { CryptoOps } from "../crypto/CryptoOps";
+import { BrowserAuthError } from "../error/BrowserAuthError";
+import { BrowserConfigurationAuthError } from "../error/BrowserConfigurationAuthError";
+import { BrowserConstants, TemporaryCacheKeys } from "../utils/BrowserConstants";
+import { BrowserStorage } from "./BrowserStorage";
+import { InMemoryStorage } from "./InMemoryStorage";
+import { IWindowStorage } from "./IWindowStorage";
+
+/**
+ * This class implements the cache storage interface for MSAL through browser local or session storage.
+ * Cookies are only used if storeAuthStateInCookie is true, and are only used for
+ * parameters such as state and nonce, generally.
+ */
+export class BrowserCacheManager extends CacheManager {
+
+    // Cache configuration, either set by user or default values.
+    private cacheConfig: CacheOptions;
+    // Window storage object (either local or sessionStorage)
+    private browserStorage: IWindowStorage;
+    // Client id of application. Used in cache keys to partition cache correctly in the case of multiple instances of MSAL.
+    private clientId: string;
+    private cryptoImpl: CryptoOps;
+
+    // Cookie life calculation (hours * minutes * seconds * ms)
+    private readonly COOKIE_LIFE_MULTIPLIER = 24 * 60 * 60 * 1000;
+
+    constructor(clientId: string, cacheConfig: CacheOptions, cryptoImpl: CryptoOps) {
+        super();
+
+        this.cacheConfig = cacheConfig;
+        this.clientId = clientId;
+        this.cryptoImpl = cryptoImpl;
+
+        if (cacheConfig.cacheLocation === BrowserConstants.CACHE_LOCATION_IN_MEMORY) {
+            this.browserStorage = new InMemoryStorage();
+        } else {
+            this.browserStorage = new BrowserStorage(cacheConfig.cacheLocation);
```

Comment:
Before this we should check if localStorage/sessionStorage throws.

Reply:
Adding now, that's why this was a draft.

Reply:
done.